File: packages/js-sdk/src/regex/detectors/za/id.ts
// South African ID Number detector

import { RegexDetector } from '../../base'
import { EntityType } from '../../entities'
import { registerRegion } from '../../registry'
import { luhnChecksum } from '../../validators'

class ZaIdDetector extends RegexDetector {
  entityType = EntityType.ZA_ID
  score = 0.85
  contextKeywords = ['id number', 'identity number', 'south african id', 'sa id']
  contextRequired = true

  // 13 digits
  pattern = /\b\d{13}\b/g
  validator = luhnChecksum
}

registerRegion('za', ZaIdDetector)

export { ZaIdDetector }

File: packages/python-sdk/src/blindfold/regex/detectors/za/id.py
"""South African ID Number detector"""

import re

from ...base import RegexDetector
from ...entities import EntityType
from ...registry import register_region
from ...validators import luhn_checksum


@register_region("za")
class ZaIdDetector(RegexDetector):
    entity_type = EntityType.ZA_ID
    score = 0.85
    context_keywords = [
        "id number", "identity number", "south african id", "sa id",
    ]
    context_required = True
    context_window = 50

    # 13 digits
    pattern = re.compile(
        r"\b\d{13}\b"
    )
    validator = staticmethod(luhn_checksum)
